src/axom/quest/DistributedClosestPoint.hpp

```diff
@@ -0,0 +1,1093 @@
+                  check_mpi_err_str_buff));
+  }
+
+  return mpi_error;
+}
+
+/**
+ * \brief Orchestrates sending and receiving conduit nodes among mpi ranks
+ *
+ * \param [in] send_node rooted node to send
+ * \param [in] recv_node rooted node to receive
+ * \param [in] send_rank MPI rank to send \a send_node to
+ * \param [in] recv_rank MPI rank to receive \a recv_node from
+ * \param [in] tag tag for MPI messages
+ * \param [in] comm MPI communicator to use
+ * 
+ * Sends/receives a pair of conduit nodes (along with their schemas) from the current rank.
+ * Uses non-blocking send (isend) and blocking receives and ensures that the request
+ * objects associated with the send are finalized
+ */
+inline void send_and_recv_node(conduit::Node& send_node,
+                               conduit::Node& recv_node,
+                               int send_rank,
+                               int recv_rank,
+                               int tag,
+                               MPI_Comm comm)
+{
+  ISendRequest req;
+
+  // non-blocking send
+  isend_using_schema(send_node, send_rank, tag, comm, &req);
+
+  // blocking receive
+  conduit::relay::mpi::recv_using_schema(recv_node, recv_rank, tag, comm);
+
+  // sender blocks until receiver is done
+  MPI_Wait(&(req.mpi_request), MPI_STATUS_IGNORE);
+}
+
+}  // namespace mpi
+}  // namespace relay
+
+/**
+ * \brief Implements the DistributedClosestPoint query for a specified dimension
+ * using a provided execution policy (e.g. sequential, openmp, cuda)
+ *
+ * \tparam NDIMS The dimension of the object mesh and query points
+ */
+template <int NDIMS>
+class DistributedClosestPointImpl
+{
+public:
+  static constexpr int DIM = NDIMS;
+  using RuntimePolicy = DistributedClosestPointRuntimePolicy;
+  using PointType = primal::Point<double, DIM>;
+  using BoxType = primal::BoundingBox<double, DIM>;
+  using PointArray = axom::Array<PointType>;
+  using BoxArray = axom::Array<BoxType>;
+
+#if defined(AXOM_USE_RAJA) && defined(AXOM_USE_UMPIRE)
+  using SeqBVHTree = spin::BVH<DIM, axom::SEQ_EXEC>;
+  #ifdef AXOM_USE_OPENMP
+  using OmpBVHTree = spin::BVH<DIM, axom::OMP_EXEC>;
+  #endif
+  #ifdef AXOM_USE_CUDA
+  using CudaBVHTree = spin::BVH<DIM, axom::CUDA_EXEC<256>>;
+  #endif
+#endif
+
+private:
+  struct MinCandidate
+  {
+    /// Squared distance to query point
+    double minSqDist {numerics::floating_point_limits<double>::max()};
+    /// Index within mesh of closest element
+    int minElem {-1};
+    /// MPI rank of closest element
+    int minRank {-1};
+  };
+
+public:
+  DistributedClosestPointImpl(RuntimePolicy runtimePolicy, bool isVerbose)
+    : m_runtimePolicy(runtimePolicy)
+    , m_isVerbose(isVerbose)
+  {
+    setAllocatorID();
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &m_rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &m_nranks);
+  }
+
+  ~DistributedClosestPointImpl()
+  {
+#if defined(AXOM_USE_RAJA) && defined(AXOM_USE_UMPIRE)
+    delete m_bvh_seq;
+    m_bvh_seq = nullptr;
```

Comment:
Minor: maybe consider making these BVH instance pointers `std::unique_ptr`, and deleting the destructor?

Reply:
done